### ISSUE TITLE
Changing Icon Packs Folder does not work

### ISSUE THREAD

**Issue by namtrah** [FlorianWoelki/obsidian-icon-folder#92]
**Describe the bug**
Error happens when trying to move icon folder:

```
Uncaught (in promise) Error: ENOENT: no such file or directory, copyfile 'C:\Notes\plugins\obsidian-icon-folder\icons\font-awesome-brands\42Group.svg' -> 'C:\Notes\plugins\icons\font-awesome-brands\42Group.svg'
```
Nothing happens afterwards. No icons have been moved. and by Restart of obsidian it uses the old directory.

**To Reproduce**
1. Go to 'Settings -> Icon Folder -> Icon Packs Folder Path'
2. Save new directory (`.obsidian/icons`)
3. Click on 'Save'

**Expected behavior**
I expect the icons to be moved and Icon folder to use the new location.

**Comment by vishae** [comment]
I have a similar issue: 

1. I rename the icon packs folder path
2. Click Save
3. Close Obsidian
4. Open Obsidian
5. The icon packs folder path field shows the old path

Additionally, if I add predefined icon packs between steps 2 and 3, the folders will save to the new path but after I restart Obsidian, the plugin would not locate the icons as the folder path would have reverted to the old path.


My workaround:

1. Navigate to <vault>/.obsidian/plugins/obsidian-icon-folder/
2. Locate the data.json file
3. Open it in a text editor
4. Manually update the "iconPacksPath" value to the new folder path
5. Save the file
6. Restart Obsidian

**Comment by FlorianWoelki** [comment]
I cannot really reproduce this issue. Is this maybe related to your OS because you are using windows?

**Comment by FlorianWoelki** [comment]
> Well, I am using windows, but I hope it is not because of that. Is there anyway I can help debug, because this is very annoying, and it seems I am not the only one with the problem.

Sure you can help. Does it throw errors when you click on save?

**Comment by namtrah** [comment]
I have worked around it (like above described) in my own Vault. but I decided to try it in a Sandbox. What I noticed from the start is that the directory is different in a fresh install. In the fresh install it defaults to `.obsidian/plugins/obsidian-icon-folder/icons`, In the upgrade install (from an old version to the latest with this change) the folder is only `plugins/obsidian-icon-folder/icons`.

Now I tried to change that directory to `.obsidian/icons` to see what happens. (which is what I did previously) and I do receive an error:
```
Uncaught (in promise) Error: ENOENT: no such file or directory, copyfile 'C:\Users\namtrah\AppData\Roaming\obsidian\Obsidian Sandbox\.obsidian\plugins\obsidian-icon-folder\icons\font-awesome-brands\42-group.svg' -> 'C:\Users\namtrah\AppData\Roaming\obsidian\Obsidian Sandbox\.obsidian\icons\font-awesome-brands\42-group.svg'
```
The directories are created `.obsidian/icons/font-awesome-brands` but no file is copied.

**Comment by namtrah** [comment]
One more update. I deleted the Icon Pack using your gui 🗑️ icon and then loaded it back in and then it did copy them in the right place. 

I'd say the logic is either messed up or there is a race condition to move the files over. And Creating a new directory needs two presses of the "SAVE" button in order to create the directories and then to save to them.

**Comment by FlorianWoelki** [comment]
> One more update. I deleted the Icon Pack using your gui 🗑️ icon and then loaded it back in and then it did copy them in the right place. 
> 
> 
> 
> I'd say the logic is either messed up or there is a race condition to move the files over. And Creating a new directory needs two presses of the "SAVE" button in order to create the directories and then to save to them.
> 
> 

I've fixed some issues with loading icon packs. Maybe you can try out the newest version (>1.3.0) and check if the issue is still there?

**Comment by namtrah** [comment]
I just updated and tried to delete the remix icons (which was missing the `Group2Line` with something like 2569 icons, and then reinstalled it...

OMG! Now there is only 170 (171a second try) icons in the pack. It still isn't really working. Or have you reduced the Remix Icon count? 

Maybe a feature request would be to have a nnn / mmmm listing of installed vs, expected so that we understand there are ones missing.

**Comment by FlorianWoelki** [comment]
> I just updated and tried to delete the remix icons (which was missing the `Group2Line` with something like 2569 icons, and then reinstalled it...
> 
> 
> 
> OMG! Now there is only 170 (171a second try) icons in the pack. It still isn't really working. Or have you reduced the Remix Icon count? 
> 
> 
> 
> Maybe a feature request would be to have a nnn / mmmm listing of installed vs, expected so that we understand there are ones missing.

I cannot really reproduce this issue. Whenever I install the Remixicon pack I always get 2271 icons.

So I think the best thing would be, if you could debug it and check why things are not downloaded correctly.

**Comment by vishae** [comment]
> I just updated and tried to delete the remix icons (which was missing the `Group2Line` with something like 2569 icons, and then reinstalled it...
> 
> OMG! Now there is only 170 (171a second try) icons in the pack. It still isn't really working. Or have you reduced the Remix Icon count?
> 
> Maybe a feature request would be to have a nnn / mmmm listing of installed vs, expected so that we understand there are ones missing.

This actually happens regularly for me, Remix Icons shows 2269 on my Macbook Air and 2270 on my mobile (and some other number on my work machine which I don't have access to right now). I try to remove and reinstall it, and sometimes it fixes it on one machine and makes it worse on another machine.
I haven't found a pattern to why this happens.

**Comment by FlorianWoelki** [comment]
Why type of sync to you use?

Maybe an alternative would be to manually download the icons.

**Comment by FlorianWoelki** [comment]
> I have worked around it (like above described) in my own Vault. but I decided to try it in a Sandbox. What I noticed from the start is that the directory is different in a fresh install. In the fresh install it defaults to `.obsidian/plugins/obsidian-icon-folder/icons`, In the upgrade install (from an old version to the latest with this change) the folder is only `plugins/obsidian-icon-folder/icons`.
> 
> Now I tried to change that directory to `.obsidian/icons` to see what happens. (which is what I did previously) and I do receive an error:
> 
> ```
> Uncaught (in promise) Error: ENOENT: no such file or directory, copyfile 'C:\Users\namtrah\AppData\Roaming\obsidian\Obsidian Sandbox\.obsidian\plugins\obsidian-icon-folder\icons\font-awesome-brands\42-group.svg' -> 'C:\Users\namtrah\AppData\Roaming\obsidian\Obsidian Sandbox\.obsidian\icons\font-awesome-brands\42-group.svg'
> ```
> 
> The directories are created `.obsidian/icons/font-awesome-brands` but no file is copied.

This issue should be resolved with the latest commit. It's related to #85

**Comment by namtrah** [comment]
I believe I have the latest version (1.3.0?) and now as I said it is even worse. I do not "sync" anything. It is a local folder. I will try again to remove the directory and reinstall it using your gui.

Can you tell me how to debug your script? Is there a way to turn on extra debug info in Obsidian? I am not a TypeScript expert.

**Comment by namtrah** [comment]
I found the actual file you download from the Remix Icons and downloaded it myself. It has all the icons in there, but they are not being copied into your directory at all on Windows 11. Only 171 now, sometimes 170 of them. I can make a list if you want and attach it. Maybe a new file is breaking your rename script
[remix-icons.zip.txt](https://github.com/FlorianWoelki/obsidian-icon-folder/files/9175508/remix-icons.zip.txt)
[remix-icons.dir.txt](https://github.com/FlorianWoelki/obsidian-icon-folder/files/9175509/remix-icons.dir.txt)
?

**Comment by FlorianWoelki** [comment]
You can check the console for warnings or issues. Make sure to have the latest version which is `1.3.2`.
So that means, that you do not have any syncing service like Obsidian Sync or something?

**Comment by namtrah** [comment]
Correct I have no syncing device at all. I do not use that (neither Obs Sync, nor OwnCloud nor MS or iCloud) I will update to 1.3.2 and test.

**Comment by FlorianWoelki** [comment]
> Correct I have no syncing device at all. I do not use that (neither Obs Sync, nor OwnCloud nor MS or iCloud) I had checked. I _am_ using 1.3.2. So it is still broken in mine.

So if there are no errors or warnings in the console, you could checkout the repository and debug it by yourself by using `console.log` statements. I think you can start debugging starting on this line:
https://github.com/FlorianWoelki/obsidian-icon-folder/blob/main/src/iconPackBrowserModal.ts#L50
There, you can make sure that all the files are read and loaded correctly.

One more question: is it just not working when you try to download the predefined icon packs? And is the remix icon pack the only one affected on?

**Comment by namtrah** [comment]
Well, I cannot define a link to download other icon packs, so yes, only the predefined ones and it happens with them all, most often with FAS and RI.

OK, to debug. Are you saying I just download the entire GitHub repository and overwrite it in Obsidian `.obsidian/plugins/obsidian-icon-folder` directory? Or is there some other program I need to debug with?

**Comment by namtrah** [comment]
```
plugin:obsidian-icon-folder:1613 Uncaught (in promise) TypeError: Cannot read properties of null (reading 'querySelector')
    at IconPackBrowserModal.eval (plugin:obsidian-icon-folder:1613:62)
    at Generator.next (<anonymous>)
    at fulfilled (plugin:obsidian-icon-folder:28:58)
eval @ plugin:obsidian-icon-folder:1613
fulfilled @ plugin:obsidian-icon-folder:28
load (async)
content @ plugin:obsidian-icon-folder:854
eval @ plugin:obsidian-icon-folder:851
eval @ plugin:obsidian-icon-folder:31
__awaiter @ plugin:obsidian-icon-folder:27
readFileSync @ plugin:obsidian-icon-folder:850
eval @ plugin:obsidian-icon-folder:1603
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
Promise.then (async)
step @ plugin:obsidian-icon-folder:30
fulfilled @ plugin:obsidian-icon-folder:28
```

**Comment by namtrah** [comment]
Could it be that the Remix Icons are no longer in a flat file system Zipped together? In all the other icon packs you search for the right directory to unpack them...

But in your definition of Remix you do not add the directories properly. Something is breaking your search...

**Comment by FlorianWoelki** [comment]
Thanks for the error.
It seems like the error occurs here:

```js
const container = this.plugin.app.workspace.containerEl.querySelector(`[data-path="${path}"]`);
const existingIconEl = container.querySelector('.obsidian-icon-folder-icon') as HTMLElement;

insertIconToNode(this.plugin, iconName, existingIconEl);
```

But that should still download all the icons.



> Could it be that the Remix Icons are no longer in a flat file system Zipped together? In all the other icon packs you search for the right directory to unpack them...
> 
> But in your definition of Remix you do not add the directories properly. Something is breaking your search...

The code goes through all the directories. So this should be not the issue.

**Comment by namtrah** [comment]
No they are no longer in a flat file.

**Comment by namtrah** [comment]
If I read your code correctly, you do not "download" each file separately rather the entire Zip file and then unpack it in Obsidian and copy over each one after normalizing the file name for your usage. 

I cannot get the console.log() messages to work. So I am of little use now. Downloading your actual repo won't work either because I do not have a TS debugger other than Obsidian itself. Sorry. Would continue if I could.

**Comment by namtrah** [comment]
Just restarted and Ouila! My logs now work. But before them I see the following errors:

![image](https://user-images.githubusercontent.com/13903735/180655846-1e5f47cb-9b99-4725-87c4-b46c0f0e5df5.png)

Which I do not understand because none of your icons are prefixed with that iconpack name (Ri, or Fas)...Does that help?

**Comment by namtrah** [comment]
I also see the following with my logs..: 
```
plugin:obsidian-icon-folder:255 loadedIconPacks
plugin:obsidian-icon-folder:262 Found 458 icons from undefined
```

**Comment by namtrah** [comment]
Sorry for the SPAM, but I am getting farther. Attached is a text file of the output of my doings. What I see is that it does read 2271 icons, but only installs 171 of them. It seems to die after community-line.svg is added.


[obsidian-1658679100466.txt](https://github.com/FlorianWoelki/obsidian-icon-folder/files/9176129/obsidian-1658679100466.txt)

**Comment by FlorianWoelki** [comment]
> Sorry for the SPAM, but I am getting farther. Attached is a text file of the output of my doings. What I see is that it does read 2271 icons, but only installs 171 of them. It seems to die after community-line.svg is added.
> 
> 
> 
> 
> 
> [obsidian-1658679100466.txt](https://github.com/FlorianWoelki/obsidian-icon-folder/files/9176129/obsidian-1658679100466.txt)
> 
> 
> 
> 

No worries!
Beautiful to hear that the logs are working for you. As I look at the error, it seems to me that your data might be corrupt in any case? The error basically appears because you might have linked a file/folder to this icon? And somehow the file/folder does not exist anymore?

**Comment by namtrah** [comment]
Well, that is because it exists in the Zip file but the extraction process of the plugin doesn't work to unpack them all.

**Comment by namtrah** [comment]
What do you know. I think you have it. I deleted my data.json file elements with the broken directories/files and then deleted the iconpack and reinstalled. And 2271 icons were installed.

There is in fact an issue when you try to somehow stop importing the icons because maybe has a lock on the broken file name?

I think this is why so many people complain about this but it is totally unrecreateable. Everyone of us have different files we deleted by hand.

**Comment by FlorianWoelki** [comment]
> What do you know. I think you have it. I deleted my data.json file elements with the broken directories/files and then deleted the iconpack and reinstalled. And 2271 icons were installed.
> 
> 
> 
> There is in fact an issue when you try to somehow stop importing the icons because maybe has a lock on the broken file name?
> 
> 
> 
> I think this is why so many people complain about this but it is totally unrecreateable. Everyone of us have different files we deleted by hand.

Thanks for investigating! I'll create an extra check to "fix" this issue.

**Comment by FlorianWoelki** [comment]
Should be fixed with [version 1.3.3](https://github.com/FlorianWoelki/obsidian-icon-folder/releases/tag/1.3.3).
